crml/nft/src/lib.rs

```diff
@@ -0,0 +1,651 @@
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin, system = frame_system {
+		type Error = Error<T>;
+
+		fn deposit_event() = default;
+
+		/// Check and close all expired listings
+		fn on_initialize(now: T::BlockNumber) -> Weight {
+			if ListingEndSchedule::<T>::contains_key(now) {
+				let removed_count = Self::close_listings_at(now);
+				// 'direct_purchase' weight is comparable to succesful closure of an auction
+				T::WeightInfo::direct_purchase() * removed_count as Weight
+			} else {
+				Zero::zero()
+			}
+		}
+
+		/// Create a new NFT collection
+		/// The caller will be come the collection' owner
+		/// `collection_id`- 32 byte utf-8 string
+		/// `schema` - for the collection
+		/// `royalties_schedule` - defacto royalties plan for secondary sales, this will apply to all tokens in the collection by default.
+		#[weight = T::WeightInfo::create_collection()]
+		fn create_collection(origin, collection_id: CollectionId, schema: NFTSchema, royalties_schedule: Option<RoyaltiesSchedule<T::AccountId>>) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+
+			ensure!(!collection_id.is_empty() && collection_id.len() <= MAX_COLLECTION_ID_LENGTH as usize, Error::<T>::CollectionIdInvalid);
+			ensure!(core::str::from_utf8(&collection_id).is_ok(), Error::<T>::CollectionIdInvalid);
+			ensure!(!<CollectionOwner<T>>::contains_key(&collection_id), Error::<T>::CollectionIdExists);
+
+			ensure!(!schema.is_empty(), Error::<T>::SchemaEmpty);
+			ensure!(schema.len() <= MAX_SCHEMA_FIELDS as usize, Error::<T>::SchemaMaxAttributes);
+
+			// Check the provided attribute types are valid
+			ensure!(
+				schema.iter().all(|(_name, type_id)| NFTAttributeValue::is_valid_type_id(*type_id)),
+				Error::<T>::SchemaInvalid
+			);
+
+			// Attribute names must be unique (future proofing for map lookups etc.)
+			let (attribute_names, _): (Vec<NFTAttributeName>, Vec<NFTAttributeTypeId>) = schema.iter().cloned().unzip();
+			let deduped = BTreeSet::from_iter(attribute_names);
+			ensure!(deduped.len() == schema.len(), Error::<T>::SchenmaDuplicateAttribute);
+
+			// Create the collection, update ownership, and bookkeeping
+			if let Some(royalties_schedule) = royalties_schedule {
+				ensure!(royalties_schedule.validate(), Error::<T>::RoyaltiesOvercommitment);
+				<CollectionRoyalties<T>>::insert(&collection_id, royalties_schedule);
+			}
+			CollectionSchema::insert(&collection_id, schema);
+			<CollectionOwner<T>>::insert(&collection_id, origin.clone());
+
+			Self::deposit_event(RawEvent::CreateCollection(collection_id, origin));
+
+			Ok(())
+		}
+
+		/// Issue a new NFT
+		/// `owner` - the token owner
+		/// `attributes` - initial values according to the NFT collection/schema
+		/// `royalties_schedule` - optional royalty schedule for secondary sales of _this_ token, defaults to the collection config
+		/// Caller must be the collection owner
+		#[weight = T::WeightInfo::create_token()]
+		#[transactional]
```

Comment:
I notice that some extrinsics and functions in this module marked as transactional so others don't. What have been the criteria for this?

Reply:
when you can't setup an 'ensure then store' situation then we need to ensure the changes rollback on failure